moc_openstack_api_reporting.py

```diff
@@ -504,51 +546,168 @@ def process_compute_events(openstack_conn, script_datetime, openstack_data, serv
+        "display_name": "vol-c8162412",
+        "event_type": "volume.create.end",
+        "generated": "2018-04-19T17:14:43.787253",
+        "host": "rbd:volumes@rbd#rbd",
+        "message_id": "63578c74-4fc5-498e-982f-e6326491a47f",
+        "project_id": "4aeb007a4f9020333a1a1be224bef276",
+        "project_name": "zealous",
+        "raw": {},
+        "request_id": "req-60bf8d5d-e80b-4748-9b5f-737674cf9428",
+        "resource_id": "64a8ece5-6853-40d6-8e56-ef919eea37f4",
+        "service": "volume.rbd:volumes@rbd#rbd",
+        "size": "9",
+        "status": "available",
+        "tenant_id": "4aeb007a4f9020333a1a1be224bef276",
+        "type": "d947c745-3a60-4ddb-abc6-97fddfe82846",
+        "user_id": "3abcb51ff942a45b52ac90915ef4c7fb",
+        "user_name": "yerwa",
+        "domain": "Default"
+    }
+    """
+
+    volume_data = openstack_data["volume_dict"][vol_id]
+    host_name = getattr(volume_data, "os-val-host-attr:host", "")
+    tenant_id = getattr(volume_data, "os-vol-tenant-attr:tenant_id", "")
+    project_rec = openstack_data["project_dict"][tenant_id]
+    volume_event_template = {
+        "availability_zone": volume_data.availability_zone,
+        "created_at": volume_data.created_at,
+        "display_name": volume_data.name,
+        "event_type": event_type,
+        "generated": volume_data.updated_at,  # will be filled out below
+        "host": host_name,
+        "message_id": "",
+        "project_id": tenant_id,
+        "project_name": project_rec["name"],
+        "raw": {},
+        "request_id": "",
+        "resource_id": volume_data.id,
+        "service": None,  # need to
+        "size": volume_data.size,
+        "status": volume_data.status,
+        "tenant_id": tenant_id,
+        "type": "",
+        "user_id": volume_data.user_id,
+        "user_name": openstack_data["user_dict"][volume_data.user_id]["name"],
+        "domain": project_rec["domain_id"],
+    }
+
+    event_to_status = {
+        "volume.create.start": "creating",
+        "volume.create.end": "available",
+        "volume.attach.start": "",
+        "volume.attach.end": "",
+        "volume.delete.start": "",
+        "volume.delete.end": "",
+    }
+
+    if volume_event_template["event_type"].startswith("volume.create"):
+        volume_event_template["generated"] = volume_data.created_at
+
+    event_list = convert_to_ceilometer_event_types(volume_event_template)
+    for ceilometer_event in event_list:
+        ceilometer_event["status"] = event_to_status[ceilometer_event["event_type"]]
+
+    return event_list
+
+
+def process_volume_events(openstack_data, cluster_state):
+    """collects and processes event data"""
+    events_by_date = {}
+    # so that we can tell if the vol not present
+    for vol_id in cluster_state["vol_timestamps"]:
+        cluster_state["vol_timestamps"][vol_id]["updated"] = 0
+
+    volume_dict = openstack_data["volume_dict"]
+    events_by_date = None
```

Comment:
I'll have to figure out which one I will keep.  This didn't cause an issue when there are volume events to report on.  It also reflects on how my thinking has evolved to handle the differences between cinder and nova.  

I am gradually convincing myself that I should be pulling the event queue despite how messy that is.

Reply:
> I'll have to figure out which one I will keep.

What do callers expect this function to return? I would probably vote for dropping line 662, so that this function is guaranteed to always return a dictionary.

```suggestion
````